lib/rules/no-duplicate-id.js

```diff
@@ -0,0 +1,139 @@
+'use strict';
+
+const NodeMatcher = require('../helpers/node-matcher');
+const Rule = require('./base');
+
+const ERROR_MESSAGE = 'ID attribute values must be unique';
+module.exports = class NoDuplicateId extends Rule {
+  logNode({ node, message }) {
+    return this.log({
+      message,
+      line: node.loc && node.loc.start.line,
+      column: node.loc && node.loc.start.column,
+      source: this.sourceForNode(node),
+    });
+  }
+  visitor() {
+    // Visiting function node filter for AttrNodes: attribute name, node type
+    function isValidIdAttrNode(node) {
+      // consider removing `@id` eventually (or make it a toggle available via config)
+      let isValidAttrNodeIdTag = ['id', '@id'].includes(node.name);
+      let isValidAttrNodeIdValueType = [
+        'TextNode',
+        'MustacheStatement',
+        'ConcatStatement',
+      ].includes(node.value.type);
+      return node && isValidAttrNodeIdTag && isValidAttrNodeIdValueType;
+    }
+
+    // Resolve MustacheStatement value to StringLiteral where possible
+    function getMustacheValue(part, scope) {
+      let refNodeStr = { type: 'MustacheStatement', path: { type: 'StringLiteral' } };
+      let refNodeDyn = { type: 'MustacheStatement', path: { type: 'PathExpression' } };
+      if (NodeMatcher.match(part, refNodeStr)) {
+        return part.path.value;
+      }
+      if (NodeMatcher.match(part, refNodeDyn)) {
+        return scope.sourceForNode(part); // this
+      }
```

Comment:
We should think about adding support for special conditionals to make this matching easier.

```js
NodeMatcher.match(part, { 
  type: 'MustacheStatement', 
  path: NodeMatcher.or([ { type: 'StringLiteral' }, { type: 'PathExpression' }])
});
```

^ would be nice here